### PR TITLE
chore(telemetry): report tested integration versions

### PR DESCRIPTION
**Pull request by wconti27**
## Motivation

With this PR we want to add a method to the generated patch for all contribs that will emit a telemetry payload describing the tested integration and its tested version. The goal of this PR is to get a list of all integrations tested during CI, along with the range of versions tested, which will be reported to metabase. The reasoning for adding another testcase instead of simply using information available in metabase from customer telemetry on what integrations are used is as follows:
 - we want the ability to know if an integration @ a specific version is being supported and actively tested, but not in use by any customers; using customer telemetry would therefore not cover this case and not allow us to deprecate support for unused integrations. 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

**Review by mabdinur — COMMENTED** [review]
Although this approach currently works it uses the instrumentation telemetry client in an unsupported way. To support python3.12 we recently changed when/how telemtry events are sent. This change will prevent integrations will not be sent in test runs: https://github.com/DataDog/dd-trace-py/pull/6859. We can hack around this issue but I don't think this is the best approach. 


Can we get the tested integration versions by parsing the venvs in https://github.com/DataDog/dd-trace-py/blob/1.x/riotfile.py? This way we can collect and submit the integration data to the test agent once. We can also make this nightly job on the 1.x branch so this data is not collected on every ci run.

File: docker-compose.yml [hunk 1]
```diff
@@ -129,13 +129,13 @@ services:
...
     testagent:
-        image: ghcr.io/datadog/dd-apm-test-agent/ddapm-test-agent:v1.11.0
+        image: docker.io/williamconti549/dd-apm-test-agent:integration-versions
         ports:
             - "127.0.0.1:9126:8126"
         volumes:
             - ./tests/snapshots:/snapshots
         environment:
-            - LOG_LEVEL=WARNING
+            - LOG_LEVEL=ERROR
```
Comment (mabdinur) [inline docker-compose.yml:138]: Why do we need to change the log level? What warnings are we seeing?

File: .circleci/config.templ.yml [hunk 2]
```diff
@@ -770,6 +776,9 @@ jobs:
 
   pytest:
     <<: *contrib_job
+    docker:
+      - image: *ddtrace_dev_image
```
Comment (mabdinur) [inline .circleci/config.templ.yml:780]: Why is this needed?

File: tests/contrib/patch.py [hunk 3]
```diff
@@ -737,3 +769,25 @@ def patch_wrapper(wrapped, _, args, kwrags):
...
 
                 self.assertEqual(out, b"OK", "stderr:\n%s" % err.decode())
+
+        def test_and_emit_get_version(self):
+            """Each contrib module should implement a get_version() function. This function is used for
+            the APM Telemetry integrations payload event, and by APM analytics to inform of dd-trace-py
+            current supported integration versions.
+            """
+            # import pdb
+            # breakpoint()
```
Comment (mabdinur) [inline tests/contrib/patch.py:779]: Make sure to remove this

File: tests/contrib/psycopg/test_psycopg_patch_generated.py [hunk 4]
```diff
@@ -36,4 +39,8 @@ def assert_module_implements_get_version(self):
         versions = get_versions()
         assert "psycopg" in versions
         assert versions["psycopg"] != ""
```
Comment (mabdinur) [inline tests/contrib/psycopg/test_psycopg_patch_generated.py:41]: nit:
```suggestion
        assert versions.get("psycopg")
```

**Review by mabdinur — COMMENTED** [review]
This is beyond the scope of this PR but some of the modules have the name `test_integrationname_patch_generated`. We need to [remove](https://github.com/DataDog/dd-trace-py/blob/v2.1.0rc1/scripts/contrib-patch-tests.py#L58) the generated part of the name and make sure the patch test is actually making the expected assertions.

File: docker-compose.yml (hunk 1 above)
Comment (wconti27) [reply docker-compose.yml]: Reduce noise during debugging but will be switched back

File: .circleci/config.templ.yml (hunk 2 above)
Comment (wconti27) [reply .circleci/config.templ.yml:788]: this adds the test agent to the job, it originally wasn't being used for pytest

File: tests/contrib/patch.py (hunk 3 above)
Comment (wconti27) [reply tests/contrib/patch.py:779]: removed

File: ddtrace/contrib/elasticsearch/patch.py [hunk 5]
```diff
@@ -50,7 +50,9 @@ def _es_modules():
     )
     for module_name in module_names:
         try:
-            yield import_module(module_name)
+            module = import_module(module_name)
+            versions[module.__name__] = getattr(module, "__versionstr__", "")
```
Comment (mabdinur) [inline ddtrace/contrib/elasticsearch/patch.py:54]:
```suggestion
            versions[module_name] = getattr(module, "__versionstr__", "")
```

File: docker-compose.yml [hunk 6]
```diff
@@ -131,13 +131,13 @@ services:
...
     testagent:
-        image: ghcr.io/datadog/dd-apm-test-agent/ddapm-test-agent:v1.11.0
+        image: ghcr.io/datadog/dd-apm-test-agent/ddapm-test-agent:v1.13.1
         ports:
             - "127.0.0.1:9126:8126"
         volumes:
             - ./tests/snapshots:/snapshots
         environment:
-            - LOG_LEVEL=WARNING
+            - LOG_LEVEL=DEBUG
```
Comment (mabdinur) [inline docker-compose.yml:140]:
```suggestion
            - LOG_LEVEL=WARNING
```

File: tests/contrib/aiohttp/test_aiohttp_patch_generated.py [hunk 7]
```diff
@@ -19,6 +19,7 @@ class TestAiohttpPatch(PatchTestCase.Base):
     __module_name__ = "aiohttp"
```
Comment (mabdinur) [inline tests/contrib/aiohttp/test_aiohttp_patch_generated.py:19]: We can do this in a follow up but the file name should be `tests/contrib/aiohttp/test_aiohttp_patch.py` (remove the _generated.py)

**Review by majorgreys — APPROVED** [review]
